### PR TITLE
fix oss client download bug; support oss recursive download;

### PR DESCRIPTION
**Pull request by bigerous**
Signed-off-by: cuidajun.cdj <cuidajun.cdj@alibaba-inc.com>

## Description
fix oss source_client download bug
refactor some dfget recursive download code:
- add URLEntry interface
- RousourceLIster.List will return []URLenty. Dfget will recursively handle entries;
- remove recursive level check method; instead decrease cfg.recursiveLevel when url represent Directory 

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

**Review by 1037husterljx — APPROVED** [review]

File: pkg/source/clients/ossprotocol/oss_source_client.go [hunk 1]
```diff
@@ -61,6 +64,7 @@ func adaptor(request *source.Request) *source.Request {
 		clonedRequest.Header.Set(oss.HTTPHeaderEtag, request.Header.Get(source.ETag))
 		clonedRequest.Header.Del(source.ETag)
 	}
+	clonedRequest.URL.Path = removeLeadingSlash(clonedRequest.URL.Path)
```
Comment (gaius-qi) [inline pkg/source/clients/ossprotocol/oss_source_client.go:67]: `strings.TrimPrefix(lonedRequest.URL.Path, string(os.PathSeparator))`

File: pkg/source/clients/ossprotocol/oss_source_client.go [hunk 2]
```diff
@@ -202,6 +209,69 @@ func (osc *ossSourceClient) GetLastModified(request *source.Request) (int64, err
...
+	delimiter := "/"
+	for {
+		lsRes, err := bucket.ListObjects(prefix, marker, oss.Delimiter(delimiter))
+		if err != nil {
+			return urls, fmt.Errorf("list oss object %s/%s: %w", request.URL.Host, path, err)
+		}
+		for _, object := range lsRes.Objects {
+			if object.Key != lsRes.Prefix {
+				url := *request.URL
+				url.Path = addLeadingSlash(object.Key)
```
Comment (gaius-qi) [inline pkg/source/clients/ossprotocol/oss_source_client.go:234]: `filepath.Join(string(os.PathSeparator), object.Key)`

File: pkg/source/clients/ossprotocol/oss_source_client.go [hunk 3]
```diff
@@ -233,11 +303,36 @@ func buildClientKey(endpoint, accessKeyID, accessKeySecret string) string {
...
+}
+
+func addLeadingSlash(s string) string {
+	if s[0] == '/' {
+		return s
+	}
+	return "/" + s
+}
+
+func addTailingSlash(s string) string {
```
Comment (gaius-qi) [inline pkg/source/clients/ossprotocol/oss_source_client.go:333]:
```golang
func addTrailingSlash(s string) string {
  if strings.HasSuffix(s, string(os.PathSeparator)) {
   return s
  }
   return s + string(os.PathSeparator)
}
```

File: pkg/source/clients/ossprotocol/oss_source_client.go [hunk 4]
```diff
@@ -202,6 +209,69 @@ func (osc *ossSourceClient) GetLastModified(request *source.Request) (int64, err
...
+	bucket, err := client.Bucket(request.URL.Host)
+	if err != nil {
+		return false, fmt.Errorf("get oss bucket %s: %w", request.URL.Host, err)
+	}
+	path := addTailingSlash(removeLeadingSlash(request.URL.Path))
+	lsRes, err := bucket.ListObjects(oss.Prefix(path), oss.Marker(""), oss.Delimiter("/"), oss.MaxKeys(1))
+	if err != nil {
+		return false, fmt.Errorf("list oss object %s/%s: %w", request.URL.Host, path, err)
+	}
+	for _, object := range lsRes.Objects {
```
Comment (gaius-qi) [inline pkg/source/clients/ossprotocol/oss_source_client.go:267]:
```golang
if len(lsRes.CommonPrefixes)+len(lsRes.Objects) > 0 {
  return true, nil
}
```

File: pkg/source/clients/ossprotocol/oss_source_client.go [hunk 5]
```diff
@@ -20,7 +20,10 @@ import (
 	"errors"
 	"fmt"
 	"net/http"
+	"net/textproto"
+	"net/url"
 	"strconv"
+	gostrings "strings"
```
Comment (gaius-qi) [inline pkg/source/clients/ossprotocol/oss_source_client.go:26]:
```golang
import (
  "strings"

  pkgstrings "d7y.io/dragonfly/v2/pkg/strings"
)

```

**Review by gaius-qi — CHANGES_REQUESTED** [review]

File: pkg/source/clients/ossprotocol/oss_source_client.go (hunk 2 above)
Comment (bigerous) [reply pkg/source/clients/ossprotocol/oss_source_client.go:234]: url里的分割符和操作系统的filepath没有关系，都统一是'/'吧

File: pkg/source/url_entry.go [hunk 6]
```diff
@@ -0,0 +1,44 @@
+package source
```
Comment (gaius-qi) [inline pkg/source/url_entry.go:1]: Remove this package and replace it with struct.

**Review by gaius-qi — CHANGES_REQUESTED** [review]

File: pkg/source/url_entry.go (hunk 6 above)
Comment (bigerous) [reply pkg/source/url_entry.go]: removed

**Review by gaius-qi — APPROVED** [review]
LGTM

File: client/config/dfget.go [hunk 7]
```diff
@@ -234,16 +234,9 @@ func (cfg *ClientOption) checkOutput() error {
...
-		if err == nil && !f.IsDir() {
-			return fmt.Errorf("path[%s] is file but requires directory path", cfg.Output)
-		}
-	} else {
-		// when not recursive download, need a file
-		if err == nil && f.IsDir() {
-			return fmt.Errorf("path[%s] is directory but requires file path", cfg.Output)
-		}
+	// when not recursive download, need a file
+	if !cfg.Recursive && err == nil && f.IsDir() {
```
Comment (pioneerlfn) [inline client/config/dfget.go:238]: In my opnion, here we shouldn't return an error, user's will feel frustated
Just use the orignal file name.

Comment (bigerous) [reply client/config/dfget.go:238]: User's dfget download request will be handled by following methed first.  
https://github.com/dragonflyoss/Dragonfly2/blob/7439b0e6c57d810b51ad0ba575b217ea88c78402/client/config/dfget.go#L171
If user not specify output with '-O' flag, then the output will be added according to the url

**Review by 1037husterljx — APPROVED** [review]

**Pull request merged**